Firestore/Source/API/FIRListenerRegistration.mm

```diff
@@ -15,46 +15,29 @@
  */
 
 #include <memory>
+#include <utility>
 
 #import "Firestore/Source/API/FIRListenerRegistration+Internal.h"
 
-#import "Firestore/Source/Core/FSTFirestoreClient.h"
-#import "Firestore/Source/Util/FSTAsyncQueryListener.h"
+#include "Firestore/core/src/firebase/firestore/util/delayed_constructor.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@implementation FSTListenerRegistration {
-  /** The client that was used to register this listen. */
-  FSTFirestoreClient *_client;
-
-  /** The async listener that is used to mute events synchronously. */
-  FSTAsyncQueryListener *_asyncListener;
+using firebase::firestore::util::DelayedConstructor;
 
-  /** The internal QueryListener that can be used to unlisten the query. */
-  std::weak_ptr<QueryListener> _internalListener;
+@implementation FSTListenerRegistration {
+  DelayedConstructor<ListenerRegistration> _registration;
```

Comment:
I'm really sorry I haven't thought about it before... It seems like `DelayedConstructor` does a strict subset of what `std::optional`/`absl::optional` does (AFAICT, `optional` doesn't require the template type to be default constructible). I don't have a strong preference here, so I'll leave it to your judgment. It seems that:
* the pro of using `optional` is having less code to maintain;
* the pro of using `DelayedConstructible` is that it's self-documenting in the sense that all instances can be easily grepped and deleted later on (one workaround is to make `DelayedConstructible` an alias for `optional`).

Reply:
I'll look into this but separately from this series. Patch stacks are nasty on github :-(.